src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1127,12 +1127,23 @@ private static bool IsUnboundGeneric(MethodSymbol method)
             return method.IsGenericMethod && method.ConstructedFrom() == method;
         }
 
+        // Arbitrary limit on the number of candidate methods for
+        // binding argument types for error recovery. If there are
+        // more candidates than this, no candidates are considered.
+        private const int MaxCandidatesForErrorRecovery = 4;
+
         private ImmutableArray<BoundExpression> BuildArgumentsForErrorRecovery(AnalyzedArguments analyzedArguments, ImmutableArray<MethodSymbol> methods)
         {
             var parameterListList = ArrayBuilder<ImmutableArray<ParameterSymbol>>.GetInstance();
-            foreach (var m in methods)
+            if (methods.Length <= MaxCandidatesForErrorRecovery)
```

Comment:
> if (methods.Length <= MaxCandidatesForErrorRecovery) [](start = 11, length = 53)

Consider applying this restriction to the final content in parameterListList rather than to an input array.